### PR TITLE
Add support for multi device foreach ops

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116064

Fix for https://github.com/pytorch/pytorch/issues/102023

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -660,6 +660,57 @@ def run_and_return_new_tensor_of_input_device(fake_mode, func, args, kwargs):
...
+            and len(arg)
+            and isinstance(arg[0], torch.Tensor)
+        ):
+            tensor_lists.append(arg)
+
+    try:
+        with in_kernel_invocation_manager(fake_mode):
+            out_meta = func(*args, **kwargs)
+    except NotImplementedError as not_implemented_error:
+        return NotImplemented
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:688]: cc @janeyx99 , @mlazos not all of the foreach ops have associated meta impls. this can be bad for memory

**Review by mlazos — APPROVED** [review]

File: test/inductor/test_foreach.py [hunk 2]
```diff
@@ -679,6 +679,23 @@ def fn(a0, a1, b0, b1):
 
         self.assertEqual(torch._inductor.metrics.generated_kernel_count, 1)
 
+    @requires_cuda()
```
Comment (mlazos) [inline test/inductor/test_foreach.py:682]: can you check that the generated kernel count is 2

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (janeyx99) [reply torch/_subclasses/fake_tensor.py:688]: @mlazos Is there already a tracker regarding adding meta impls for foreach ops? It seems like a good idea to add them.

**Pull request closed without merging**